contrib/win32/win32compat/ssh-agent/connection.c

```diff
@@ -136,26 +159,94 @@ get_con_client_type(struct agent_connection* con)
 	    GetTokenInformation(token, TokenUser, info, info_len, &info_len) == FALSE)
 		goto done;
 
-	sid_size = SECURITY_MAX_SID_SIZE;
-	if (CreateWellKnownSid(WinLocalSystemSid, NULL, system_sid, &sid_size) == FALSE)
-		goto done;
-	sid_size = SECURITY_MAX_SID_SIZE;
-	if (CreateWellKnownSid(WinNetworkServiceSid, NULL, ns_sid, &sid_size) == FALSE)
-		goto done;
-	sid_size = SECURITY_MAX_SID_SIZE;
-	if (CreateWellKnownSid(WinLocalServiceSid, NULL, ls_sid, &sid_size) == FALSE)
-		goto done;
+	/* check if its localsystem */
+	{
+		sid_size = SECURITY_MAX_SID_SIZE;
+		if (CreateWellKnownSid(WinLocalSystemSid, NULL, sid, &sid_size) == FALSE)
+			goto done;
+		if (EqualSid(info->User.Sid, sid)) {
+			con->client_type = SYSTEM;
+			r = 0;
+			goto done;
+		}
+			
+	}
+
+	/* check if its SSHD service */
+	{
+		/* Does NT Service/SSHD exist */
+		LookupAccountNameW(NULL, sshd_act, NULL, &sshd_sid_len, NULL, &reg_dom_len, &nuse);
+		
+		if (GetLastError() == ERROR_NONE_MAPPED)
+			debug3("Cannot look up SSHD account, its likely not installed");
+		else if (GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
+			error("LookupAccountNameW on SSHD account failed with %d", GetLastError());
+			goto done;
+		} else {
+			if ((sshd_sid = malloc(sshd_sid_len)) == NULL ||
+			    (ref_dom = (wchar_t*)malloc(reg_dom_len * 2)) == NULL ||
+			    LookupAccountNameW(NULL, sshd_act, sshd_sid, &sshd_sid_len, ref_dom, &reg_dom_len, &nuse) == FALSE)
+				goto done;
+
+			if (EqualSid(info->User.Sid, sshd_sid)) {
+				con->client_type = SSHD_SERVICE;
+				r = 0;
+				goto done;
+			}
+			if (CheckTokenMembership(token, sshd_sid, &isMember) == FALSE)
+				goto done;
+			if (isMember) {
+				con->client_type = SSHD_SERVICE;
+				r = 0;
+				goto done;
+			}
+		}
+	}
 
-	if (EqualSid(info->User.Sid, system_sid) ||
-	    EqualSid(info->User.Sid, ls_sid) ||
-	    EqualSid(info->User.Sid, ns_sid))
-		con->client_type = MACHINE;
-	else
-		con->client_type = USER;
+	/* check if its LS or NS */
+	{
+		sid_size = SECURITY_MAX_SID_SIZE;
+		if (CreateWellKnownSid(WinNetworkServiceSid, NULL, sid, &sid_size) == FALSE)
```

Comment:
just need to call IsWellKnownSid(info->User.Sid, WinNetworkServiceSid) to findout if it is network service. no need to create the sid and compare